# Homework 1

Link to exercise:
https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/cohorts/2025/01-intro/homework.md

## Setup

Environment: Conda `mlops-zoomcamp`

In [1]:
# dependencies
import numpy as np
import pandas as pd
import pyarrow.parquet as pq
from pathlib import Path
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
# paths
PATH_REPO = Path.cwd().parent.parent
PATH_DATA_DIR = PATH_REPO / 'data'
PATH_DATA_JAN = PATH_DATA_DIR / 'yellow_tripdata_2023-01.parquet'
PATH_DATA_FEB = PATH_DATA_DIR / 'yellow_tripdata_2023-02.parquet'


## Q1. Downloading the data

Question: How many columns are in the dataset?

The data was downloaded previously from the website linked in the exercise.


In [3]:
# load data from file and convert to pandas data frame
trips_jan = pq.read_table(PATH_DATA_JAN).to_pandas()
trips_feb = pq.read_table(PATH_DATA_FEB).to_pandas()

In [4]:
# print data, number of columns is specified below
trips_jan

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee  
0                         2.5         0.00  
1                         2.5         0.00  
2                         2.5         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
3066761                   NaN          NaN  
3066762                   NaN          NaN  
3066763                   NaN          NaN  
3066764                   N

There are 19 columns in the dataset.

## Q2. Computing duration

What's the standard deviation of the trips duration in January?

In [5]:
# compute duration column
trips_jan['duration'] = trips_jan['tpep_dropoff_datetime'] - trips_jan['tpep_pickup_datetime']

# convert to minutes
trips_jan['duration'] = trips_jan['duration'].dt.total_seconds() / 60

# compute and print result
print(
    f'Standard deviation of the trips duration in January: ',
    f'{trips_jan["duration"].std().round(2)}'
)

Standard deviation of the trips duration in January:  42.59


## Q3. Dropping outliers

What fraction of the records is left after you dropped the outliers?

In [6]:
# get number of records before dropping outliers
n_records_before = trips_jan.shape[0]

# drop outliers according to exercise
trips_jan = trips_jan[
    (trips_jan['duration'] >= 1) & (trips_jan['duration'] <= 60)
]

# get number of records after dropping outliers
n_records_after = trips_jan.shape[0]

# compute and print result
print(
    f'Fraction of the records left after dropping outliers: ',
    f'{n_records_after / n_records_before:.2f}'
)

Fraction of the records left after dropping outliers:  0.98


## Q4. One-hot encoding

What's the number of columns of the one-hot encoded matrix?

In [7]:
# select only the location columns, convert to strings and make a dictionary
locations_dict = trips_jan[
    ['PULocationID', 'DOLocationID']
].astype(str).to_dict(orient='records')

# create and fit a dict vectorizer
dv = DictVectorizer(sparse=False) # get numpy array instead of sparse matrix
feature_matrix = dv.fit_transform(locations_dict)

# get number of columns
print(f'Number of columns: {feature_matrix.shape[1]}')

Number of columns: 515


## Q5. Training a model

What's the RMSE on train?

In [8]:
# get labels
labels = trips_jan['duration']
labels

0           8.433333
1           6.316667
2          12.750000
3           9.616667
4          10.833333
             ...    
3066761    13.983333
3066762    19.450000
3066763    24.516667
3066764    13.000000
3066765    14.400000
Name: duration, Length: 3009173, dtype: float64

In [12]:
# initialize the model
model = LinearRegression()

I started training the model on the full data set, which for a tutorial is
actually pretty extensive in size.
It has about 3 million rows, and 515 columns due to the one-hot encoding.
After about three hours the training was still not complete, so I decided to
stop it.
I am not sure if I got something wrong and downloaded the wrong data set.
Anyway, I will not go back now and try figuring it out.
Instead I will subset the data to 10,000 rows and train a smaller model.
Since this is a tutorial, I think this is a reasonable approach.

In [21]:
# subset the data to 10,000 rows

subset_size = 10000

feature_matrix = feature_matrix[:subset_size]
labels = labels[:subset_size]

In [22]:
# fit the model
model.fit(
    feature_matrix,
    labels
)

LinearRegression()

In [23]:
# get predictions
predictions_train = model.predict(feature_matrix)

# compute RMSE on train set
rmse_train = np.sqrt(mean_squared_error(labels, predictions_train))
print(f"RMSE: {rmse_train:.2f} minutes")

RMSE: 7.76 minutes


## Q6. Evaluating the model

What's the RMSE on validation?

In [24]:
# subset february data, too, to save time
trips_feb = trips_feb[:subset_size]

# compute duration for February data
trips_feb['duration'] = (
    trips_feb['tpep_dropoff_datetime'] - 
    trips_feb['tpep_pickup_datetime']
).dt.total_seconds() / 60

# apply the same duration filter as for January data
trips_feb = trips_feb[
    (trips_feb['duration'] >= 1) & 
    (trips_feb['duration'] <= 60)
]

# prepare features the same way
val_dicts = trips_feb[
    ['PULocationID', 'DOLocationID']
].astype(str).to_dict(orient='records')

# transform using the SAME DictVectorizer that was fit on January data
val_feature_matrix = dv.transform(val_dicts)

# get labels
val_y = trips_feb['duration'].values

In [27]:
# get predictions for val set
predictions_val = model.predict(val_feature_matrix)

# compute RMSE on val set
rmse_val = np.sqrt(mean_squared_error(val_y, predictions_val))
print(f"RMSE: {rmse_val:.2f} minutes")

RMSE: 10.02 minutes
